index.js

```diff
@@ -250,47 +242,77 @@
       return match == null ? '' : match[1];
     };
 
+  function getStaticMethod(_name) {
+    var name = 'fantasy-land/' + _name;
+    return function(typeRep) {
+      return (
+        typeRep != null && typeof typeRep[name] === 'function' ?
+          typeRep[name] :
+        typeof typeRep === 'function' ?
+          implementations[functionName (typeRep) + '.' + name] :
+        // else
+          null
+      );
+    };
+  }
+
+  function getPrototypeMethod(_name) {
+    var name = 'fantasy-land/' + _name;
+    return function(x) {
+      return (
+        x != null &&
+        (x.constructor == null || x.constructor.prototype !== x) &&
+        typeof x[name] === 'function' ?
+          x[name] :
+        // else
+          implementations[type (x) + '#' + name]
+      );
+    };
+  }
+
   //  $ :: (String, Array TypeClass, StrMap (Array Location)) -> TypeClass
   function $(_name, dependencies, requirements) {
-    function getBoundMethod(_name) {
-      var name = 'fantasy-land/' + _name;
-      return requirements[_name] === Constructor ?
-        function(typeRep) {
-          return (funcPath ([name], typeRep)) ||
-                 (typeof typeRep === 'function'
-                  ? funcPath ([functionName (typeRep), name], implementations)
-                  : null);
-        } :
-        function(x) {
-          var isPrototype = x != null &&
-                            x.constructor != null &&
-                            x.constructor.prototype === x;
-          var m = (isPrototype ? null : funcPath ([name], x)) ||
-                  (funcPath ([type (x), 'prototype', name], implementations));
-          return m && m.bind (x);
-        };
-    }
-
     var version = '12.1.0';  // updated programmatically
-    var keys = Object.keys (requirements);
+
+    var staticMethodNames = [];
+    var prototypeMethodNames = [];
+    Object.keys (requirements)
+    .forEach (function(_name) {
+      switch (requirements[_name]) {
+        case Constructor:
+          return staticMethodNames.push (_name);
+        case Value:
+          return prototypeMethodNames.push (_name);
+      }
+    });
 
     var typeClass = TypeClass (
       'sanctuary-type-classes/' + _name,
       'https://github.com/sanctuary-js/sanctuary-type-classes/tree/v' + version
         + '#' + _name,
       dependencies,
       function(x) {
-        return keys.every (function(_name) {
-          var arg = requirements[_name] === Constructor ? x.constructor : x;
-          return getBoundMethod (_name) (arg) != null;
-        });
+        return (
+          staticMethodNames.every (function(_name) {
+            return x != null &&
+                   getStaticMethod (_name) (x.constructor) != null;
```

Comment:
The test suite passes without the `x != null` guard. All of this library's type classes that require static methods happen to depend on type classes that require prototype methods that `Null` and `Undefined` cannot provide.

These steps demonstrate that `x.constructor` should be guarded:

1.  Remove the `x != null` guard.
2.  Remove the `Semigroupoid` dependency from `Category`.
3.  Evaluate `Z.Category.test (null)`.

Since users can define their own type classes via `Z.TypeClass`, it is necessary to include the guard.

Reply:
> users can define their own type classes [which necessitate inclusion of] the guard

That sounds like it might warrant a test.